src/models/ClassPotentialN2HDM.cpp

```diff
@@ -803,93 +734,301 @@ void Class_Potential_N2HDM::TripleHiggsCouplings()
-    if (testsum > ZeroThreshold) posG2 = i;
-    testsum = std::abs(HiggsRot(i, 4)) + std::abs(HiggsRot(i, 5));
-    if (testsum > ZeroThreshold) posG0 = i;
+    // Charged submatrix
+    if (std::abs(HiggsRot(i, pos_rho1))
+        + std::abs(HiggsRot(i, pos_rho2)) > ZeroThreshold)
+    {
+      pos_G1 = i;
+    }
+    else if (std::abs(HiggsRot(i, pos_eta1))
+        + std::abs(HiggsRot(i, pos_eta2)) > ZeroThreshold)
+    {
+      pos_G2 = i;
+    }
+    // Neutral CP-odd submatrix
+    else if (std::abs(HiggsRot(i, pos_psi1))
+        + std::abs(HiggsRot(i, pos_psi2)) > ZeroThreshold)
+    {
+      pos_G0 = i;
+    }
+    else
+    {
+      throw std::runtime_error("Error. Non-Goldstone in Goldstone submatrix."
+                               " Particle with negative mass? Check your "
+                               "parameter point.");
+    }
   }
+
+  // Now going from 3 to NHiggs, i.e. fixing the physical Higgs bosons
   for (std::size_t i = 3; i < NHiggs; i++)
+    // mass base index i corresponds to mass vector sorted in ascending mass
+  {
+    // Charged submatrix
+    if (std::abs(HiggsRot(i, pos_rho1))
+        + std::abs(HiggsRot(i, pos_rho2)) > ZeroThreshold)
+    {
+      pos_H1 = i;
+    }
+    else if (std::abs(HiggsRot(i, pos_eta1))
+        + std::abs(HiggsRot(i, pos_eta2)) > ZeroThreshold)
+    {
+      pos_H2 = i;
+    }
+    // Neutral CP-odd submatrix
+    else if (std::abs(HiggsRot(i, pos_psi1))
+        + std::abs(HiggsRot(i, pos_psi2)) > ZeroThreshold)
+    {
+      pos_A = i;
+    }
+    // Neutral CP-even submatrix
+    else if (std::abs(HiggsRot(i, pos_zeta1)) + std::abs(HiggsRot(i, pos_zeta2))
+        + std::abs(HiggsRot(i, pos_rhoS)) > ZeroThreshold)
+      // use that mh1 < mh2 < mh3
+    {
+      if (pos_h1 == -1) {
+        pos_h1 = i;
+      } else if (pos_h2 == -1) {
+        pos_h2 = i;
+      } else {
+        pos_h3 = i;
+      }
+    }
+  }
+
+  // check if all position indices are set
+  if (pos_G0 == -1 or pos_G1 == -1 or pos_G2 == -1 or
+      pos_H1 == -1 or pos_H2 == -1 or
+      pos_h1 == -1 or pos_h2 == -1 or pos_h3 == -1 or pos_A == -1)
+  {
+    throw std::runtime_error("Error. Not all position indices are set.");
+  }
+
+  // check if all other elements of rotation matrix are zero
+  bool zero_element = false;
+  for (std::size_t i = 0; i < NHiggs; i++)
   {
-    testsum = std::abs(HiggsRot(i, 0)) + std::abs(HiggsRot(i, 1));
-    if (testsum > ZeroThreshold) posMHCS1 = i;
-    testsum = std::abs(HiggsRot(i, 2)) + std::abs(HiggsRot(i, 3));
-    if (testsum > ZeroThreshold) posMHCS2 = i;
-    testsum = std::abs(HiggsRot(i, 6)) + std::abs(HiggsRot(i, 7)) +
-              std::abs(HiggsRot(i, 8));
-    if (testsum > ZeroThreshold)
+    for (std::size_t j = 0; j < NHiggs; j++)
     {
-      posN.at(countposN) = i;
-      countposN++;
+      int ii = int(i);
+      int jj = int(j);
+      if (not((jj == pos_rho1  and (ii == pos_G1 or ii == pos_H1)) or
```

Comment:
This piece of code is a bit confusing, what is it checking for? Can you refactor it a little bit?